### ISSUE TITLE
Type inference bug on 0.3.5 and release-0.3

### ISSUE THREAD

**Issue by garborg** [JuliaLang/julia#9835]
Fine on 0.4.
```
type LLA
           lat::Float64
           lon::Float64
           alt::Float64
end

type ENU
           east::Float64
           north::Float64
           up::Float64
end

for pt in [ENU(-1.1, 2.2, -3.3)]
           typ = isa(pt, LLA) ? LLA : ENU
           r = typ(2.2, 3.3, 4.4)
end
```
```
ERROR: typ not defined
 in inlining_pass at ./inference.jl:2628
 in inlining_pass at ./inference.jl:2590
 in inlining_pass at ./inference.jl:2553
 in typeinf at ./inference.jl:1544
 in typeinf_ext at ./inference.jl:1216
```

**Comment by simonster** [comment]
Maybe the same issue as #9630, although your test case is a lot more useful.

**Comment by garborg** [comment]
Thanks for linking them up -- looks like I got my keywords mixed up when I searched.

**Comment by ihnorton** [comment]
Slightly simpler test-case:
```
type A end; type B end
f() = begin pt = A()
    typ = isa(pt, A) ? A : B
    typ()
end
f()
```

the error occurs here:
```
inlining_pass: 
e => :($(Expr(:call1, :(typ::Union(Type{A},Type{B})))))
```

leading to eval'ing `:(typ::Union(Type{A},Type{B}))`, but `typ` doesn't exist at that point.

Possible fix (@vtjnash?):
```
diff --git a/inference.jl b/inference.jl
index ebeb2a1..5162949 100644
--- a/inference.jl
+++ b/inference.jl
@@ -2617,6 +2617,8 @@ function inlining_pass(e::Expr, sv, ast)
             ET = exprtype(arg1)
             if isType(ET)
                 f = ET.parameters[1]
+            elseif isconstantfunc(arg1,sv) === false
+                f = abstract_eval(arg1, emptydict, sv)
             else
                 f = _ieval(arg1)
             end
```

**Comment by vtjnash** [comment]
that seems correct

**Comment by ihnorton** [comment]
fixed in #10044
